### PR TITLE
Handle shard allocation for NONE migration direction

### LINKED ISSUE #13321: [Remote Store] Handle shard allocation for NONE migration direction
### Is your feature request related to a problem? Please describe

* Before migration starts, new nodes join the cluster in `MIXED` compatibility mode, while the migration direction remains at its default value of `NONE`.
* We should handle shard allocation correctly under these conditions, based on the index settings.

### Describe the solution you'd like

The following scenarios should be handled in `RemoteStoreMigrationAllocationDecider.java`: 
* For remote store backed indices, shards should only be allocated to remote store nodes.
* For non remote store backed indices, shards should only be allocated to non-remote store or docrep nodes.

### Related component

Storage:Remote

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ltaragi**
### Description
* Before migration starts, new nodes join the cluster in `MIXED` compatibility mode, while the migration direction remains at its default value of `NONE`.
* We should handle shard allocation correctly under these conditions, based on the index settings.
* This change handles the following scenarios in `RemoteStoreMigrationAllocationDecider.java`: 
  * For remote store backed indices, shards should only be allocated to remote store nodes.
  * For non remote store backed indices, shards should only be allocated to non-remote store or docrep nodes.
* Extends validation for `TranslogDurabilitySettings` to the case of remote store migration by adding extra condition of `isMigratingToRemoteStore(clusterSettings) == false` [here](https://github.com/opensearch-project/OpenSearch/pull/13322/files#diff-ea2e29ab98d6120bb1c62077f504340d6465e5178b885631170f45d209838668R1662)

### Related Issues
Resolves #13321

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] ~New functionality has been documented.~
  - [ ] ~New functionality has javadoc added~
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [ ] ~Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))~
- [ ] ~Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/cluster/routing/allocation/decider/RemoteStoreMigrationAllocationDecider.java [hunk 1]
```diff
@@ -95,18 +95,24 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
...
             );
+        } else if (migrationDirection.equals(Direction.DOCREP)) {
+            // docrep migration direction is currently not supported
+            return allocation.decision(Decision.YES, NAME, getDecisionDetails(true, shardRouting, targetNode, " for DOCREP direction"));
         }
 
         // check for remote store backed indices
-        IndexMetadata indexMetadata = allocation.metadata().getIndexSafe(shardRouting.index());
-        boolean remoteStoreBackedIndex = IndexMetadata.INDEX_REMOTE_STORE_ENABLED_SETTING.get(indexMetadata.getSettings());
         if (remoteStoreBackedIndex && targetNode.isRemoteStoreNode() == false) {
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/decider/RemoteStoreMigrationAllocationDecider.java:116]: we can move all of the `else`  block , to enhance readability .

File: server/src/main/java/org/opensearch/cluster/routing/allocation/decider/RemoteStoreMigrationAllocationDecider.java [hunk 2]
```diff
@@ -95,18 +95,24 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
             );
         }
 
-        if (migrationDirection.equals(Direction.REMOTE_STORE) == false) {
-            // docrep migration direction is currently not supported
+        IndexMetadata indexMetadata = allocation.metadata().getIndexSafe(shardRouting.index());
+        boolean remoteStoreBackedIndex = IndexMetadata.INDEX_REMOTE_STORE_ENABLED_SETTING.get(indexMetadata.getSettings());
```
Comment (shourya035) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/decider/RemoteStoreMigrationAllocationDecider.java:99]: nit: Can we change this to `remoteSettingsBackedIndex`?

**Review by shourya035 — APPROVED** [review]

File: server/src/internalClusterTest/java/org/opensearch/remotemigration/RemoteStoreMigrationShardAllocationIT.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+import org.opensearch.cluster.routing.allocation.decider.Decision;
+import org.opensearch.test.OpenSearchIntegTestCase;
+
+import java.util.Locale;
+import java.util.Optional;
+
+import static org.opensearch.node.remotestore.RemoteStoreNodeService.CompatibilityMode.MIXED;
+
+@OpenSearchIntegTestCase.ClusterScope(scope = OpenSearchIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class RemoteStoreMigrationShardAllocationIT extends RemoteStoreMigrationShardAllocationBaseTestCase {
```
Comment (gbbafna) [inline server/src/internalClusterTest/java/org/opensearch/remotemigration/RemoteStoreMigrationShardAllocationIT.java:23]: Lets use `RemoteMigrationAllocationDeciderIT` to add this test case .

File: server/src/internalClusterTest/java/org/opensearch/remotemigration/RemoteStoreMigrationShardAllocationBaseTestCase.java [hunk 4]
```diff
@@ -110,6 +146,108 @@ protected ShardRouting getShardRouting(boolean isPrimary) {
...
+                }
+            }
+        }
+
+        assertNotNull(requiredDecision);
+        return requiredDecision;
+    }
+
+    // get allocation and relocation decisions for all nodes
+    protected void prepareDecisions() {
```
Comment (gbbafna) [inline server/src/internalClusterTest/java/org/opensearch/remotemigration/RemoteStoreMigrationShardAllocationBaseTestCase.java:193]: name doesn't make sense here. Do we need to rename to `clearExclusion` ?

File: server/src/internalClusterTest/java/org/opensearch/remotemigration/RemoteStoreMigrationShardAllocationIT.java (hunk 3 above)
Comment (gbbafna) [reply server/src/internalClusterTest/java/org/opensearch/remotemigration/RemoteStoreMigrationShardAllocationIT.java:23]: Do we need to add more test cases for `remote_store` direction , which was left out in previous PR ?

**Comment by ltaragi** [comment]
> ❕ Gradle check result for [54bfe73](https://github.com/opensearch-project/OpenSearch/commit/54bfe7314c6bcc4114e55a39d1c1deb041cad8f0): [UNSTABLE](https://build.ci.opensearch.org/job/gradle-check/38067/)
> 
> * **TEST FAILURES:**
> 
> ```
>       2 org.opensearch.common.util.concurrent.QueueResizableOpenSearchThreadPoolExecutorTests.classMethod
>       1 org.opensearch.common.util.concurrent.QueueResizableOpenSearchThreadPoolExecutorTests.testResizeQueueDown
> ```
> 
> Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

Flaky test #11979

**Review by gbbafna — APPROVED** [review]

**Pull request merged**